packages/core/src/sandbox/linux/landlock-runner.rs

```diff
@@ -0,0 +1,313 @@
+    exit(126);
+}
+
+fn normalize_path_and_access(
+    path: &OsString,
+    write: bool,
+    abi: i32,
+) -> (OsString, u64) {
+    let mut access = access_mask(write, abi);
+
+    // Landlock expects the path in a PATH_BENEATH rule to be a directory.
+    // If a caller passed a file (e.g., /dev/null), fall back to its parent
+    // directory and trim creation/removal permissions to avoid broadening
+    // privileges unnecessarily.
+    if let Ok(metadata) = std::fs::metadata(path) {
+        if metadata.is_dir() {
+            return (path.clone(), access);
+        }
+
+        // Narrow permissions when we have to allow a parent directory.
+        const REMOVE_MASK: u64 = ACCESS_FS_REMOVE_DIR | ACCESS_FS_REMOVE_FILE;
+        const MAKE_MASK: u64 = ACCESS_FS_MAKE_CHAR
+            | ACCESS_FS_MAKE_DIR
+            | ACCESS_FS_MAKE_REG
+            | ACCESS_FS_MAKE_SOCK
+            | ACCESS_FS_MAKE_FIFO
+            | ACCESS_FS_MAKE_BLOCK
+            | ACCESS_FS_MAKE_SYM;
+        access &= !(REMOVE_MASK | MAKE_MASK);
+
+        let parent: PathBuf = Path::new(path)
+            .parent()
+            .map(Path::to_path_buf)
+            .unwrap_or_else(|| PathBuf::from("/"));
+        return (parent.into_os_string(), access);
+    }
+
+    (path.clone(), access)
+}
+
+fn add_rule(fd: c_int, path: impl AsRef<std::ffi::OsStr>, access: u64) {
+    let cpath = match CString::new(path.as_ref().as_bytes()) {
+        Ok(p) => p,
+        Err(_) => {
+            eprintln!(
+                "[landlock-runner] Skipping path with interior null: {:?}",
+                path.as_ref()
+            );
+            return;
+        }
+    };
+
+    let parent_fd = unsafe { open(cpath.as_ptr(), O_PATH | O_CLOEXEC) };
+    if parent_fd < 0 {
+        eprintln!(
+          "[landlock-runner] Failed to open path {:?}: errno {}",
+          path.as_ref(),
+          last_errno()
+        );
+        exit(111);
+    }
+
+    let rule = landlock_path_beneath_attr {
+        allowed_access: access,
+        parent_fd,
+    };
+
+    let res = unsafe {
+        syscall(
+            SYS_LANDLOCK_ADD_RULE,
+            fd,
+            LANDLOCK_RULE_PATH_BENEATH,
+            &rule as *const _,
+            0 as c_int,
+        )
+    };
+
+    unsafe { close(parent_fd) };
+
+    if res != 0 {
+        eprintln!(
+            "[landlock-runner] landlock_add_rule failed for {:?}: errno {}",
+            path.as_ref(),
+            last_errno()
+        );
+        exit(111);
+    }
+}
+
+fn access_mask(write: bool, abi: i32) -> u64 {
+    let ro = ACCESS_FS_EXECUTE | ACCESS_FS_READ_FILE | ACCESS_FS_READ_DIR;
+    let mut rw = ro
+        | ACCESS_FS_WRITE_FILE
+        | ACCESS_FS_REMOVE_DIR
+        | ACCESS_FS_REMOVE_FILE
+        | ACCESS_FS_MAKE_CHAR
+        | ACCESS_FS_MAKE_DIR
+        | ACCESS_FS_MAKE_REG
+        | ACCESS_FS_MAKE_SOCK
+        | ACCESS_FS_MAKE_FIFO
+        | ACCESS_FS_MAKE_BLOCK
+        | ACCESS_FS_MAKE_SYM;
+
+    if abi >= 2 {
```

Comment:
The backwards compatibility logic is wrong about the access rights an in which ABI versions they started appearing. Please consult the Landlock documentation regarding this (or better, use the Landlock crate, if you can):

https://docs.kernel.org/userspace-api/landlock.html#defining-and-enforcing-a-security-policy

"refer" appeared in ABI v2, "truncate" in ABI v3, "ioctl-dev" in ABI v5